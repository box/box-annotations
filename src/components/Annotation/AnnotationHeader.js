// @flow
import * as React from 'react';
import { FormattedMessage } from 'react-intl';

import InlineDelete from '../../../third-party/components/InlineDelete';
import UserLink from '../../../third-party/components/UserLink';
import Timestamp from './Timestamp';
import messages from './messages';

import './AnnotationHeader.scss';

type Props = {
    id: string,
    permissions: AnnotationPermissions,
    createdAt?: string,
    createdBy?: User,
    onDelete?: Function
};

const AnnotationHeader = ({ id, permissions, onDelete, createdAt, createdBy }: Props) => {
    const annotator =
        !!createdBy && !!createdBy.name
            ? createdBy
            : {
                id: '0',
                name: <FormattedMessage className='ba-annotation-user-name' {...messages.anonymousUserName} />
            };

    return (
        <div className='ba-annotation-headline'>
            <UserLink className='ba-annotation-user-name' id={annotator.id} name={annotator.name} />
            {createdAt && <Timestamp time={createdAt} />}
            {onDelete && (
                <InlineDelete
                    id={id}
                    permissions={permissions}
                    message={<FormattedMessage {...messages.annotationDeletePrompt} />}
                    onDelete={onDelete}
                />
            )}
        </div>
    );
};

export default AnnotationHeader;
