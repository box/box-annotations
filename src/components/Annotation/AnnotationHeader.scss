@import '../../variables.scss';

.ba .ba-annotation-headline {
    display: flex;
    flex-flow: row wrap;

    &:hover,
    &.ba-is-focused {
        .bcs-comment-delete-container {
            opacity: 1;
        }
    }

    .ba-annotation-avatar {
        flex-shrink: 0;
        margin-right: 17px; // Spaces avatar circle from text to the right
        margin-top: 3px; // Required to top align avatar circle with text to the right
    }

    .ba-annotation-user-name {
        color: $twos;
        font-weight: bold;
        margin-right: 10px; // Spaces user name from timestamp
    }

    .ba-annotation-created-at {
        color: $nines;
        cursor: default;
        font-size: smaller;
    }

    .bcs-comment-delete-container {
        margin-top: 3px; // Required to top align trash icon with text to the left
        opacity: 0;
        position: absolute;
        right: 0;
        top: 0;
        z-index: 1;

        &.ba-is-confirming {
            opacity: 1;
        }
    }

    .bcs-comment-delete {
        vertical-align: top;
    }
}

.flyout-overlay {
    .bcs-comment-delete-yes,
    .bcs-comment-delete-no {
        font-size: inherit;
    }
}

.ba-mobile-annotation-dialog .ba-annotation-headline {
    .ba-annotation-avatar {
        width: 43px;

        img {
            height: 40px;
            width: 40px;
        }
    }

    .ba-annotation-user-name {
        font-size: 16px;
    }

    .bcs-comment-delete-container {
        font-size: 15px;
        opacity: 1;
    }

    .bcs-comment-delete svg {
        height: 18px;
        width: 18px;
    }
}
