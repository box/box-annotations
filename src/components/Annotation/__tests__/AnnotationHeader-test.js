import * as React from 'react';
import { shallow } from 'enzyme';

import AnnotationHeader from '../AnnotationHeader';

const TIME_STRING_SEPT_27_2017 = '2017-09-27T10:40:41-07:00';

const USER = {
    type: 'user',
    id: '789ghi',
    name: 'Mia Thermopolis',
    email: 'princess@genovia.gov'
};
const ANONYMOUS_USER = 'Some User';

const headerProps = {
    id: '123',
    createdAt: TIME_STRING_SEPT_27_2017,
    createdBy: USER,
    permissions: {}
};

describe('components/Annotation/AnnotationHeader', () => {
    const render = (props = {}) => shallow(<AnnotationHeader {...headerProps} {...props} />).dive();

    test('should correctly render annotation header', () => {
        const wrapper = render({ isAnonymousUser: true });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('Timestamp').prop('time')).toEqual(TIME_STRING_SEPT_27_2017);
        expect(wrapper.find('UserLink').prop('name')).toEqual(USER.name);
    });

    test('should correctly display anonymous username if no name was provided in createdBy', () => {
        const unknownUser = {
            type: 'user',
            id: '0'
        };
        const wrapper = render({ createdBy: unknownUser });
        expect(wrapper.find('UserLink').prop('name')).toEqual(ANONYMOUS_USER);
    });

    // eslint-disable-next-line
    test('should allow user to delete if they have delete permissions on the annotation and delete handler is defined', () => {
        const wrapper = render({ onDelete: jest.fn(), permissions: { can_delete: true } });
        expect(wrapper.find('InlineDelete').length).toEqual(1);
    });

    test('should not allow user to delete if they lack delete permissions on the annotation', () => {
        const wrapper = render({ onDelete: jest.fn() });
        expect(wrapper.find('InlineDelete').length).toEqual(0);
    });

    test('should not allow user to delete if onDelete handler is undefined', () => {
        const wrapper = render();
        expect(wrapper.find('InlineDelete').length).toEqual(0);
    });
});
