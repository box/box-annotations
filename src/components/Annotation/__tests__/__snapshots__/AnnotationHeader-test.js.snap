// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Annotation/AnnotationHeader should correctly render annotation header 1`] = `
<div
  className="ba-annotation-headline"
>
  <UserLink
    className="ba-annotation-user-name"
    id="789ghi"
    name="Mia Thermopolis"
  />
  <Timestamp
    time="2017-09-27T10:40:41-07:00"
  />
</div>
`;
